proposals/0082-relax-transaction-constraints.md

```diff
@@ -0,0 +1,173 @@
+These constraints must still be satisfied in order for the transaction to be
+executed, and have an effect on state. However, there are some new considerations
+which must be taken into account, specifically as it relates to fees and block
+limits.
+
+### Fee-Paying
+
+Currently, iff a transaction's fee-payer does not have enough funds to pay the
+fee, the transaction cannot be included in a block. With this proposal, it is
+possible for such a transaction to be included in the block, and there are
+three different edge-cases to consider:
+
+1. The fee-payer account does not exist (0 lamports)
+2. The fee-payer account does not have enough funds for the entire fee
+3. The fee-payer account has enough funds for the fee, but would no longer be
+rent-exempt
+
+In case 1, the transaction should simply be ignored for execution, and have no
+effect on state.
+In case 2, the fee-paying account should be drained of all funds, but have no
+other effect on state.
+In case 3, the fee-paying account should be drained of all funds, with fees
+being paid to the block-producer, and the remainder of lamports being dropped.
+
+### Block-Limits
+
+Pending the activation of `2ry7ygxiYURULZCrypHhveanvP5tzZ4toRwVp89oCNSj`,
+validators must validate a block is within block-limits.
+With this proposal, some transactions may not be executable, but will still
+count towards block-limits.
+If these transactions did not count towards block-limits, the validation of
+block-limits would require the validator to check whether or not a transaction
+is executable, which negates the benefits of this proposal.
+
+Additionally, f these transactions did not count towards block-limits, a
+malicious leader could produce a block with non-executable transactions and
+overload the network.
+
+## Impact
+
+- Transactions that would previously be dropped with an error, can now be
+  included, and even charged fees.
+  - Users must be more careful when constructing transactions to ensure they
+    are executable if they don't want to waste fees
+- The validity of a block is no longer dependent on intra-block account-state
+updates. This is because the only account-state required for transaction
+validation is the ALT resolution, which is resolved at the beginning of a slot.
+  - Block-production could be done asynchronously
+  - Block-validation could be done without execution, but still relies on the
+  execution of previous blocks.
+
+## Security Considerations
+
+- Removing fee-paying requirement could allow validator clients to produce
+  blocks that do not pay fees. If not checked by block-producer, this could
+  allow malicious users to abuse such producers/leaders.
+
+## Drawbacks
+
+- Any dynamic fees based on block utilization cannot/should not reward the
+block-producer directly. Otherwise, this incentivizes block-producers to fill
+blocks with non-fee paying transactions in order to raise fees.
+- Non-fee paying transactions included in a block will be recorded in long-term
```

Comment:
This could be a way to attack the cluster and use more space. A malicious leader could create 48M CU / (CU required to load payer account which currently stands at 0) number of transactions. This could be a massive number of transactions, depending on CUs charge that we decide to load the payer account. If each transaction was given the maximum size of 1280 bytes or eventually more if we include ZK transactions, then I can possibly create a huge block burdening the whole cluster's network resources because of the turbine.

I feel like the leader should be somehow punished/slashed for adding transactions where the payer does not have enough balance.

Reply:
Currently transactions do have a minimum cost of 1070 CUs = 720 SigVerify + 300 WriteLock.
This corresponds to a maximum of 44776 txs per slot, assuming 48M CU limit.

Additionally, cost has 1 CU per 4 bytes of instruction data. So large ix data txs would cost more than this bare minimum of 1070 CUs as well.
It's still a large number of transactions of 34532 txs per slot.

Reply:
It's worth noting that currently, all direct rewards are going to the block-producer anyway.
If that large number of transactions were valid fee-paying transactions, it would only be more work for other validators than if they could not pay fees. Either way, other nodes are not directly rewarded for the validation/execution.

There is still some discussion to be had around **indirect** rewards via the burning of SOL in those fees. I think that's potentially a valid reason we may want to force the block-producer into paying fees. Hoping to discuss this more

Reply:
Yeah I agree, block producers should pay a fee as they are using network resources, storage space to transfer invalid transactions. Malicious node can just start creating blocks with invalid transactions of 1280 bytes using 44.20096 MBs/ block in network and storage.